Support backwards compatible operation by default

pixelb opened the pull request:
Add the "version_base" parameter to hydra.main,
which acts as central way to select the defaults for that version.

Without a specified "version_base" the user will get a warning,
and we'll default to the currently supported minimum version defaults
(currently those of version 1.1).

Specifying a "version_base" will suffice to select appropriate defaults,
rather than requiring one explicitly set other config items like
config_path, or hydra.job.chdir.

New hydra uses should use version_base=None to get
the default set of the currently installed version.

pixelb committed:
support backwards compatible operation by default

Add the "version_base" parameter to hydra.main,
which acts as central way to select the defaults for that version.

Without a specified "version_base" the user will get a warning,
and we'll default to the currently supported minimum version defaults
(currently those of version 1.1).

Specifying a "version_base" will suffice to select appropriate defaults,
rather than requiring one explicitly set other config items like
config_path, or hydra.job.chdir.

New hydra uses should use version_base=None to get
the default set of the currently installed version.

Jasha10 commented:
Let me outline an argument in favor of using hydra's [`Singleton`](https://github.com/facebookresearch/hydra/blob/1aa46fbe460c7e3d136687d9ae5cda7f98aecfbb/hydra/core/singleton.py#L8) class to encapsulate the version base (instead of using `global __version_base__`):

Remote launcher plugins need to serialize Hydra's global state so that the remote process that will run the job can use the same global state. For example, the Hydra submitit launcher uses this pattern:
  - [create a list](https://github.com/facebookresearch/hydra/blob/1aa46fbe460c7e3d136687d9ae5cda7f98aecfbb/plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py#L131) of jobs to launch, [including the `Singleton` state](https://github.com/facebookresearch/hydra/blob/1aa46fbe460c7e3d136687d9ae5cda7f98aecfbb/plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py#L141) as an input for these jobs
  - call the submitit executor with this list of jobs
  - in each remote process corresponding to one of the jobs, [call `Singleton.set_state`](https://github.com/facebookresearch/hydra/blob/1aa46fbe460c7e3d136687d9ae5cda7f98aecfbb/plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py#L60) before composing the config for that specific job.

Since the final config is composed on the remote process, that remote process must have access to Hydra's global state. I believe that this should include the version base, as the composition process will probably depend on the version base (e.g. PR #2056 removes deprecated functionality that affects config composition, and so the version base will play a role in finalizing that PR). The `Singleton.get_state` and `Singleton.set_state` methods allow simultaneously serializing or restoring the states of all of Hydra's singleton subclasses (e.g. `ConfigStore`, `Plugins`, `GlobalHydra`, etc) using a single access point.

pixelb committed:
change from config_path=None default to version_base=None default

  $ git grep -l hydra.main.*config_path=None | grep -Ev '(docs/upgrades|versioned_docs)' |
    xargs sed -i 's/hydra.main(config_path=None/hydra.main(version_base=None/'

  $ git grep -l initialize.*config_path=None | grep -Ev '(docs/upgrades|versioned_docs)' |
    xargs sed -i 's/initialize(config_path=None)/initialize(version_base=None)/'

pixelb committed:
add version_base=None to initialize() and hydra.main() in tests etc.

  $ git grep hydra.main.*config_path= | grep -v version_base | grep -Ev '(docs/upgrades|versioned_docs)' |
    cut -d: -f1 | uniq | xargs sed -i 's/hydra\.main(config_path=/hydra.main(version_base=None, config_path=/'

Also various manual additions

pixelb committed:
version_base: tests

jieru-hu commented on examples/advanced/defaults_list_interpolation/my_app.py:
```diff
@@ -4,7 +4,7 @@
 
 
-@hydra.main(config_path="conf", config_name="config")
+@hydra.main(version_base=None, config_path="conf", config_name="config")
```
Thank you! overall it looks great!
My question is do we need to show `version_base=None` in all the examples? For new users starting afresh they probably won't need to know this till later (there is fair amount to ramp up on when first starting Hydra, so IMO less is better). 
For existing users, they are likely to change this version_base at upgrade time and we will include the pointer to version_base in the upgrade guide.

pixelb replied:
Note that an unspecified version_base currently gives a warning.
Also I was removing settings of config_path=None as that's implied with version_base=None.
Also I didn't add version_base params to most of the intro docs for the reasons you state.
I'll push an update with docs, to hopefully clarify this more.
Happy to adjust after that

pixelb committed:
default to hydra.job.chdir=False with version_base >= "1.2"

Also adjust docs to remove the implication we'll break behavior in hydra 1.3

pixelb committed:
default to config_path=None with version_base >= "1.2"

Update the website/docs/upgrades/*/hydra_main_config_path.md docs,
to indicate the new version_base default setting for config_path.

pixelb committed:
version_base: docs

Jasha10 commented on website/docs/upgrades/version_base.md:
```diff
@@ -0,0 +1,19 @@
+---
+id: version_base
+title: version_base description
```
```suggestion
title: version_base
```

Jasha10 commented on website/docs/upgrades/version_base.md:
```diff
@@ -0,0 +1,19 @@
+---
+id: version_base
+title: version_base description
+---
```
It would be good to add `'upgrades/version_base'` and `'upgrades/1.1_to_1.2/hydra_main_config_path'` to the `website/sidebars.js` file so that these new markdown files can be found via the hydra docs sidebar.

jieru-hu commented:
lg! should we add some notes on version_base in the developer guide as well? This is helpful to know  for introducing breaking changes.

Jasha10 reviewed (approved):
Looks good to me!
